# Bayesian Sequential Subtree approach - part 1

## 1. `BASEML`

To calculate the branch lengths, the Hessian, and the gradient required for `MCMCtree` to approximate the likelihood calculation, we first need to run `BASEML`. In directory [`02_SeqBayes_S2/01_BASEML`](../../02_SeqBayes_S2/01_BASEML/README.md), you will find a `README.md` file with very detailed instructions about the steps we followed for this analysis:

* [Prepare the file structure to run `BASEML` on an HPC cluster](../../02_SeqBayes_S2/01_BASEML/README.md#1-prepare-architecture-and-prepare-baseml-files): if you have gone through this repository, you will see that all the pipelines, code snippets, and various tools work under a specific file structure with the aim to improve data reproducibility. In the first part of the [aforementioned `README.md` file](../../01_SeqBayes_S1/01_BASEML/02_Hessian/README.md#1-loading-data-to-the-hpc-and-set-working-environment), you will learn how to set up the file structure we used to run `BASEML`.
* [Generate a directory for the final `in.BV` files](../../02_SeqBayes_S2/01_BASEML/README.md#2-generate-directory-for-5-partitions-alignment): we do not go into more details with regards to how to run `BASEML` as we had already written a thorough explanation when carrying out the first part of the Bayesian Sequential Subtree approach (see [`README.md` file](../../01_SeqBayes_S1/01_BASEML/02_Hessian/README.md)). To this end, the last part of [the aforementioned `README.md` file](../../02_SeqBayes_S2/01_BASEML/README.md) focuses on explaining how we generated the `in.BV` files for each data subset (i.e., as explained [in the step-by-step tutorial](../../02_SeqBayes_S2/README.md#2-second-part-of-the-sequential-bayesian-dating-approach), the second data subset was further divided into 13 data subsets according to taxonomical clades).

We have provided relevant input/output files which size is less than 100Mb (file size limit to be uploaded on a GitHub repository) in the following subdirectories:

* [`Afrotheria`](00_BASEML/Afrotheria/)
* [`Euarchonta`](00_BASEML/Euarchonta/)
* [`Lagomorpha`](00_BASEML/Lagomorpha/)
* [`Laurasiatheria_cetartiodactyla`](00_BASEML/Laurasiatheria_cetartiodactyla/)
* [`Laurasiatheria_chiroptera_subt1`](00_BASEML/Laurasiatheria_chiroptera_subt1/)
* [`Laurasiatheria_chiroptera_subt2`](00_BASEML/Laurasiatheria_chiroptera_sub21/)
* [`Laurasiatheria_therest`](00_BASEML/Laurasiatheria_therest/)
* [`Marsupialia`](00_BASEML/Marsupialia/)
* [`Rodentia_ctenohystrica`](00_BASEML/Rodentia_ctenohystrica/)
* [`Rodentia_squirrel`](00_BASEML/Rodentia_squirrel/)
* [`Rodentia_subtree1`](00_BASEML/Rodentia_subtree1/)
* [`Rodentia_subtree2`](00_BASEML/Rodentia_subtree2/)
* [`Xenarthra`](00_BASEML/Xenarthra/)

Inside of each directory (one per data subset) you shall find the following subdirectories:

* `1`: input tree file (`tmp0001.trees`), control file (`tmp0001.ctl`), and `rst2` output file when using the concatenated alignment block with the 5 partitions we had generated (i.e., mitochondrial 1st+2nd codon positions, mitochondrial 3rd codon positions, mRNA, nuclear 1st+2nd codon positions, and nuclear 3rd codon positions).
* `2`: input tree file (`tmp0001.trees`), control file (`tmp0001.ctl`), and `rst2` output file when using the mitochondrial alignment with 1st and 2nd codon positions (i.e., `mt_12cp`).
* `3`: input tree file (`tmp0001.trees`), control file (`tmp0001.ctl`), and `rst2` output file when using the mitochondrial alignment with 3rd codon positions (i.e., `mt_3cp`).
* `4`: input tree file (`tmp0001.trees`), control file (`tmp0001.ctl`), and `rst2` output file when using the mRNA alignment (i.e., `mt_rna`).
* `5`: input tree file (`tmp0001.trees`), control file (`tmp0001.ctl`), and `rst2` output file when using the nuclear alignment with 1st and 2nd codon positions (i.e., `nt_12cp`).
* `6`: input tree file (`tmp0001.trees`), control file (`tmp0001.ctl`), and `rst2` output file when using the nuclear alignment with 3rd codon positions (i.e., `nt_3cp`).
* `7`: a file called `rst2` with the concatenated blocks saved in the `rst2` files output for each of the five alignment blocks (partitions) in directories `2` to `6` in the same order: `mt_12cp`, `mt_3cp`, `mt_rna`, `nt_12cp`, and `nt_3cp`.

>> **TIP** you can run command `grep '^ [0-9]' */rst2` from each data subset directory to check the number of taxa present in each alignment block in directories `1` to `7`.
>> **NOTE 1**: please note that we did not include the `tmp0001.txt` files despite being less than 100Mb to avoid increasing the size of this GitHub repository -- and make it easier and faster to clone! You can download all the input/output files from [our Dropbox archive](https://www.dropbox.com/scl/fi/31koa6sj985up03213f8c/SeqBayesS2_BASEML.zip?rlkey=farhdhykszi9kto78nx5qykao&st=4hc69ar8&dl=0).
>> **NOTE 2**: please read the `README.txt` files for the following subdirectories regarding the `rst2` files:
>>
>> * `Euarchonta/[1-6]`
>> * `Laurasiatheria_chiroptera_subt1/1`
>> * `Laurasiatheria_chiroptera_subt2/1`
>> * `Laurasiatheria_therest/1`
>> * `Rodentia_subtree1/1`
>> * `Rodentia_subtree2/1`

## 2. `MCMCtree`

Once we had generated the `in.BV` files for each tree hypothesis, we were ready to run `MCMCtree`. In directory [`02_SeqBayes_S2`](../../02_SeqBayes_S2/README.md), you will find a `README.md` file with the following sections:

* [Summary of the analyses carried out during the first part of the Bayesian Sequential Subtree approach](../../02_SeqBayes_S2/README.md#1-quick-summary-of-tasks-accomplished-in-step-1): we provide an overview of all the steps we followed to obtain the backbone tree when using the first data subset during our Bayesian Sequential Subtree approach. We also detail the procedure we followed to [fit skew-t (ST) and skew-normal (SN) distributions to the posterior time densities for all the nodes of our backbone tree](../../02_SeqBayes_S2/README.md#13-fitting-skew-t-distributions-to-resulting-posterior-densities).
* [Overview of the second data subset and the main analyses that were to be carried out](../../02_SeqBayes_S2/README.md#21-outline): all the details regarding how we further subset our second data subset and the alignment blocks we analysed are given in this section.
* [Quick summary to run `BASEML`](../../02_SeqBayes_S2/README.md#22-run-baseml-to-estimate-the-hessian-and-the-gradient): short section that summarises the key points of this analysis as [further details were given during the first part of the Bayesian Sequential Subtree approach](../../01_SeqBayes_S1/01_BASEML/02_Hessian/README.md).
* [Quick summary to run `MCMCtree`](../../02_SeqBayes_S2/README.md#23-run-mcmctree-to-infer-species-divergence-times-for-each-data-subset): short section that summarises the key points of timetree inference. We also redirect the readers to directory [`00_data_curation`](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/), where they will find one `README.md` file for each data subset (i.e., `00_data_curation/<name_datasubset>/filter_aln/README.md` for sequence alignment and `00_data_curation/<name_datasubset>/filter_tree_updchr/README.md` for subtree) detailing all the filtering steps we followed to parse and curate these datasets.

A summary of our analyses when running `MCMCtree` when sampling from the prior and the posterior can be found below:

* [Prior](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation):
  * Given that we used both fossil information and ST and SN distributions to constrain the node ages of our 13 subtrees, we ran various iterations to check that the node age constraints based on ST/SN distributions were not in conflict with those based on the fossil record. You can find all the details regarding these checks if you (i) navigate to the [`00_data_curation` directory](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/), (ii) access each of the directories there (i.e., one per data subset/subtree). and (iii) then read the corresponding `README.md` files under `filter_tree_updchr`. You will find all the information you need to reproduce all our analyses. We have not uploaded all the files used/generated during the checks here so that the size of this GitHub repository is manageable. Please download the analyses for each data subset from our Dropbox archive when sampling from the prior:
    * [Afrotheria - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/afrotheria/filter_tree_updcrh/README.md): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/sbu212y8oeiaqwz/SeqBayesS2_check_conflict_afrotheria.zip?st=vqgosu0x&dl=0).
    * [Laurasiatheria chiroptera - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/chiroptera/filter_tree_updcrh/README.md): you can download the results from these checks [for the first chiroptera subtree](https://www.dropbox.com/s/fnzbelxhzknwqr2/SeqBayesS2_check_conflict_chirosubt1.zip?st=gzmpoypl&dl=0) and [for the second chiroptera subtree](https://www.dropbox.com/s/soqn03bm8gcoar3/SeqBayesS2_check_conflict_chirosubt2.zip?st=ogpjths0&dl=0) from our Dropbox archive.
    * [Euarchonta - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/euarchonta/filter_tree_updcrh/README.md): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/l17ct03llwhrttj/SeqBayesS2_check_conflict_euarchonta.zip?st=35hdibc7&dl=0).
    * [Lagomorpha - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/lagomorpha/filter_tree_updcrh/README.md): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/50lsxk3gw9b1gr8/SeqBayesS2_check_conflict_lagomorpha.zip?st=qnqh0sdk&dl=0).
    * [Laurasiatheria cetartiodactyla - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/laurasiatheria_cetartiodactyla/filter_tree_updcrh/README.md):[for version 1 (soft-bound calibration for Whippomorpha)](https://www.dropbox.com/s/ngwgzfm0thsqsim/SeqBayesS2_check_conflict_artiodactyla.zip?st=hmvp0xbd&dl=0) [and version 2 (lower-bound calibration for Whippomorpha)](https://www.dropbox.com/s/xr68808m1vqnmnz/SeqBayesS2_check_conflict_artiodactyla_v2.zip?st=7gqjzfgv&dl=0).
    * [Laurasiatheria the rest - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/laurasiatheria_therest/filter_tree_updcrh/README.md): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/frurexfq2ntyxu9/SeqBayesS2_check_conflict_laurasiatheria_therest.zip?st=5xy3dape&dl=0).
    * [Marsupialia - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/marsupialia/filter_tree_updcrh/README.md): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/n7zm9mt73jg2jwl/SeqBayesS2_check_conflict_marsupialia.zip?st=2grm3e81&dl=0).
    * [Rodentia ctenohystrica - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/rodentia_ctenohystrica/): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/yuwld0r7saeo1m9/SeqBayesS2_check_conflict_ctenohystrica.zip?st=rlzf249y&dl=0).
    * [Rodentia squirrel - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/rodentia_squirrel/filter_tree_updcrh/README.md): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/lm0fii20l5taz5k/SeqBayesS2_check_conflict_sciuridae.zip?st=6xafq9f9&dl=0).
    * [Rodentia the rest - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/afrotheria/filter_tree_updcrh/README.md): you can download the results from these checks [for the first rodentia subtree](https://www.dropbox.com/s/tu9mcwel50953c6/SeqBayesS2_check_conflict_rodsubt1.zip?st=zzjq9xhy&dl=0) and [the second rodentia subtree](https://www.dropbox.com/s/8yglaevkid1zrsq/SeqBayesS2_check_conflict_rodsubt2.zip?st=9jzhpkl5&dl=0) from our Dropbox archive.
    * [Xenarthra - assess marginal densities VS calibration densities and potential truncation issues](../../02_SeqBayes_S2/00_Data_filtering/00_data_curation/xenarthra/filter_tree_updcrh/README.md): you can download the results from these checks [from our Dropbox archive](https://www.dropbox.com/s/6o3qpwp1mu4y25w/SeqBayesS2_check_conflict_xenarthra.zip?st=d012al5y&dl=0).
* [Posterior](01_MCMCtree): please note that we did not upload `mcmc.txt` files (i.e., output files by `MCMCtree` with all the samples collected) so that this repository does not take a long time to be cloned (we ran 32 chains for each of the 13 data subsets; the original compressed file is ~12Gb!). You can download such files from [our Dropbox archive](https://www.dropbox.com/s/6gb746s7usitmcl/SeqBayesS2_MCMCtree.zip?st=10c2h5rh&dl=0):
  * **Independent chains, directories `7/run[1-32]`**: if you have checked the details about [our file structure for the `BASEML` analyses above](README.md#1-baseml), you will see that the directory for each data subset had a subdirectory named `7`, where the `in.BV` file with the 5 `rst2` blocks for partitions `mt_12cp`, `mt_3cp`, `mt_rna`,`nt_12cp`, and `nt_3cp` had been concatenated in this order (i.e., same order as in the input sequence file!). We followed the same file structure for timetree inference with `MCMCtree`, and thus each directory under [`01_MCMCtree/<name_datasubset>`] has a subdirectory named `7`, where the analyses took place. Inside `7`, there are 32 subdirectories `run[1-32]/mcmctree_GBM`, one for each chain we ran. There, you will find the control file, the `in.BV` file, and the input tree file that we used for each individual run. You will see that some of the `<name_datasubsets>` directories have one tree file inside `<name_datasubsets>` while others have a duplicate of the input tree for each `run[1-32]/mcmctree_GBM`. When we downloaded the data from the HPC cluster we used, the soft links we created inside `run[1-32]/mcmctree_GBM` would not redirect to the main tree file, and thus we had to delete them. In such cases, you will only see the main tree file in directory `7`. While these files are not too large, we cannot upload the full file structure or, otherwise, directory `01_MCMCtree` would have ~9Gb worth of data. Therefore, we have decided to just upload the final timetree inferred for each data subset, the seed numbers we used for all the chains we ran for each data subset (i.e., 32 chains per data subset; 32 seed numbers), the input tree file, an example of control file to run `MCMCtree` (i.e., this control file was generated during our pipeline and modified the paths to the input files automatically), and a comma-separated file named `Calibs_nodes<name_dataubset>.csv`, which contains a summary of the node age constraints applied to the corresponding subtree (3 columns: node name, node number as per `MCMCtree`, constraint used). The `in.BV` files can be found in the corresponding data subset directories in [`00_BASEML`](00_BASEML).
  * **Merging independent chains**: to merge those chains that had converged to the same target distributions, we used our in-house script `Combine_MCMC.sh`, which you shall find inside `<name_datasubset>/7/` if you download the timetree inference files from [our Dropbox archive](https://www.dropbox.com/s/6gb746s7usitmcl/SeqBayesS2_MCMCtree.zip?st=10c2h5rh&dl=0). If you navigate to `<name_datasubset>/7/mcmc_files/` (and `<name_datasubset>/7/mcmc_files_filt/` if not all the chains were used to summarise the results; see `Combine_MCMC_filt.sh` script too), you shall find the final timetrees for each data subset, which are named `FigTree.tre`. There, you shall also find the control file that we used with option `print = -1` to obtain the final timetrees. Here, we have included the bash scripts we used to merge the chains.
